fixes 4381: dns records are being update after changes to host when dhcp is also present.

dmitri-d opened the pull request.

dmitri-d committed:
fixes #4381: dns records are being update after changes to host when dhcp is also present.

domcleal commented:
@lzap could you review please?

lzap commented:
Damn, I was hoping for Docker one :-) Sure thing.

domcleal commented:
@lzap take as many as you like/can ;)

lzap commented:
I was able to test, it works fine now.

lzap commented on app/models/concerns/orchestration/dns.rb:
```diff
@@ -119,8 +119,8 @@ def dns_conflict_detected?
-    status &= failure(_("DNS PTR Records %s already exists") % dns_ptr_record.conflicts.to_sentence, nil, :conflict) if reverse_dns? and dns_ptr_record and dns_ptr_record.conflicting?
-    status
+    status = failure(_("DNS PTR Records %s already exists") % dns_ptr_record.conflicts.to_sentence, nil, :conflict) if reverse_dns? and dns_ptr_record and dns_ptr_record.conflicting?
+    not status #failure method returns 'false'
```
The failure method returns always _false_, so why not to put _true_ here as a constant?

domcleal replied:
@lzap only if we failed.. the default is `status = true` on line 120

lzap replied:
Ah okay.

lzap commented:
:+1:

domcleal commented:
So a few comments, but I'm unsure which of these, if any, we should fix in this PR or consider separate issues.  The whole area seems fraught with bugs IMO.
1. I get a DHCP conflict warning when I'm changing a hostname, which seems odd: `DHCP records conflicttest5.example.com-52:54:00:de:7a:09/192.168.122.111 already exists` (conflicttest5 is the _current_ hostname, which I was changing from.)  I get them in other places too, not introduced by this patch.
2. Introduced by this is an error if you change the IP of a host: `Net::Conflict in HostsController#update in DHCP detected - expected conflicttest5.example.com-52:54:00:de:7a:09/192.168.122.111, found conflicttest5.example.com-52:54:00:de:7a:09/192.168.122.111`.  I can't tell what's happening here, it's confused about itself and breaks half way through orchestration.
3. When changing IP on `develop`, it doesn't seem to update DNS records.
4. Minor one, "#" is missing from the commit msg.

dmitri-d commented:
1: I, too, find this behaviour weird. I decided to preserve it however in order to minimize the changes in this PR. 
2 & 3: I'll take a look, I think those should be fixed in this PR.

dmitri-d commented:
@domcleal: fixed the issue you discovered. The fix makes an assumption that we don't want to change leases for hosts not under Foreman's control. I added a new error dialog to advise users that they should resolve dhcp lease conflicts before proceeding.

lzap commented:
I will re-test now focusing more on DHCP (and DNS as well).

lzap commented:
Tested scenarios:
- Renamed to existing DNS A entry by changing host name. PASS
- Renamed to existing DNS A entry by changing IP address. NOT_RELEVANT
- Renamed to existing DHCP entry by changing MAC address. PASS
- Renamed to existing DHCP entry by changing IP address. PASS

I can confirm (1) - it is weird and also the format of the message is confusing (and not relevant in this case):

```
DHCP records rename1.virtual.lan-45:5e:c0:e9:dd:a5/192.168.100.16 already exists
```

I guess more clear would be (for relevant cases):

```
the following DHCP records already exists:
 - 45:5e:c0:e9:dd:a5/192.168.100.16 (rename1.virtual.lan) 
```

The same applies for DNS when hostname is not touched. Again, out of topic for this PR.

Couple of other issues (not sure if I should file them - not relevant):
- When I click on the ACK dialog anywhere, the default behavior is "acknowledge". But not big deal, there is still Overwrite and Cancel button.
- When I hit the Cancel button, 404 is generated because the target path is renamed host (but we cancelled that).

Is there anything else I can test?

dmitri-d commented:
@lzap: looks quite exhaustive. I'll check the 404 error.

lzap commented:
I don't think the cancel routing error was caused by your patch. +1 from my side

domcleal commented on app/views/hosts/_form.html.erb:
```diff
@@ -123,6 +124,6 @@
 
     <%= f.hidden_field :overwrite? %>
-    <%= submit_or_cancel f, @host.overwrite?, :cancel_path => @host %>
+    <%= submit_or_cancel f, @host.overwrite?, :cancel_path => edit_host_path(@host.name_changed? ? @host.name_was : @host.name) %>
```
This line appears to be causing test failures.

domcleal replied:
@witlessbird?

dmitri-d replied:
Will retest/update today (was waiting for fix for #4851 to get merged).

domcleal replied:
Ok, thanks.  (I still see these test failures.)

dmitri-d commented:
@domcleal: updated & rebased.

dmitri-d commented:
@domcleal: updated & rebased for realz now.

domcleal commented:
I've had a fun morning testing this :unamused:

There's a bug that affects normal provisioning - if you have a DHCP-enabled host, you can't cancel or set the build mode, as a DHCP conflict is detected.

```
Failed to enable errors7.example.com for installation: ["Conflict DHCP records errors7.example.com-52:54:00:44:82:a0/192.168.122.121 already exists"]
```

I've tracked this down to the `conflicts` method in lib/net/dhcp/record.rb which fails to exclude itself from the records coming back from the smart proxy (the delete_if should do the trick, but doesn't).  This wasn't a problem on develop as the `dhcp_conflict_detected?` method was erroneously returning false (this line: https://github.com/theforeman/foreman/pull/1293/files#diff-49b1433f3c6173eeb125e4a223f13a8bL142).

Maybe we should override `==` on Net::DHCP::Record?

dmitri-d commented:
@domcleal: changed the equality on Net::DHCP::Record

domcleal commented:
Looks good and works well, thanks!

domcleal commented:
Merged as 90ff2a3 for Foreman 1.5.0, thanks @witlessbird!
